### PR TITLE
feat: add missed governance votes monitor

### PR DESCRIPTION
**Pull request by NeverHappened**
Think it's good enough, but look at the logic of getting validators that missed votes please.
And I look at the addresses, not the `valoper` node addresses, cause it was simpler (API types don't have it).
Is it ok, or better to get `valoper` operator addresses?

### CODE REVIEW

File: docker/config/grafana/provisioning/dashboards/lido_bots/lido_bots.json [hunk 1]
```diff
@@ -319,7 +319,7 @@
         ],
         "executionErrorState": "alerting",
         "for": "5m",
-        "frequency": "1m",
+        "frequency": "24h",
```
Comment (oopcode) [inline docker/config/grafana/provisioning/dashboards/lido_bots/lido_bots.json:322]: nice

File: docker/env/.lido_terra.testnet.env [hunk 2]
```diff
@@ -11,4 +11,5 @@ ADDRESSES_REWARDS_DISPATCHER_CONTRACT=terra16c4qpu6xg2l399dr7nf2xcuscqkg22pcnn5z
 GRAFANA_PORT=3000
 SERVICE_NAME=lido_terra_testnet
 DELEGATIONS_DISTRIBUTION_CONFIG_NUM_MEDIAN_ABSOLUTE_DEVIATIONS=3
-NETWORK_GENERATION=columbus-5
\ No newline at end of file
+NETWORK_GENERATION=columbus-5
+MISSING_VOTES_GOVERNANCE_MONITOR_MONITORED_VALIDATORS=terra1v5hrqlv8dqgzvy0pwzqzg0gxy899rm4kdn0jp4,terra123gn6j23lmexu0qx5qhmgxgunmjcqsx8g5ueq2,terra15zcjduavxc5mkp8qcqs9eyhwlqwdlrzy6anwpg,terra1v5hrqlv8dqgzvy0pwzqzg0gxy899rm4kdn0jp4,terra1kprce6kc08a6l03gzzh99hfpazfjeczfpd6td0,terra1c9ye54e3pzwm3e0zpdlel6pnavrj9qqvqgf7ps,terra144l7c3uph5a7h62xd8u5et3rqvj3dqtvve3he0,terra1542ek7muegmm806akl0lam5vlqlph7spfs99vq,terra1sym8gyehrdsm03vdc44rg9sflg8zeuqwfd3384,terra1khfcg09plqw84jxy5e7fj6ag4s2r9wqsg5jt4x,terra15urq2dtp9qce4fyc85m6upwm9xul30496lytpd,terra1alpf6snw2d76kkwjv3dp4l7pcl6cn9uytq89zk,terra1nwrksgv2vuadma8ygs8rhwffu2ygk4j24pxxx0,terra175hhkyxmkp8hf2zrzka7cnn7lk6mudtv4nsp2x,terra13g7z3qq6f00qww3u4mpcs3xw5jhqwraswv3q3t,terra1jkqr2vfg4krfd4zwmsf7elfj07cjuzss3qsmhm,terra15cupwhpnxhgylxa8n4ufyvux05xu864jcrrkqa
```
Comment (oopcode) [inline docker/env/.lido_terra.testnet.env:15]: isn't it better to query the validators registry contract directly? https://finder.terra.money/columbus-5/address/terra16j67029v236npntxjatwena6kh7rgvkwdkw0tm

```
{"get_validators_for_delegation": {}}
```

File: internal/app/collector/monitors/missing_votes_governance_monitor.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+func (m *MissingVotesGovernanceMonitor) FetchNotVotedValidators(ctx context.Context, proposalID int) ([]string, error) {
+	res := make([]string, 0)
+
+	proposalVotes, err := m.repository.GetVotes(ctx, proposalID)
+	if err != nil {
+		return res, fmt.Errorf("failed to get votes with {%s}", err)
+	}
+
+	votedValidatorsSubset := make([]string, 0)
```
Comment (sotnikov-s) [inline internal/app/collector/monitors/missing_votes_governance_monitor.go:159]: `votedValidatorsSubset := make([]string, 0, len(proposalVotes))`
when the result slice size is known, it's a good practice the allocate the slice of that size

File: internal/app/collector/monitors/missing_votes_governance_monitor.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+	return result, nil
+}
+
+func (m *MissingVotesGovernanceMonitor) FetchNotVotedValidators(ctx context.Context, proposalID int) ([]string, error) {
+	res := make([]string, 0)
+
+	proposalVotes, err := m.repository.GetVotes(ctx, proposalID)
+	if err != nil {
+		return res, fmt.Errorf("failed to get votes with {%s}", err)
```
Comment (sotnikov-s) [inline internal/app/collector/monitors/missing_votes_governance_monitor.go:156]: it's idiomatic to return nil in error responses, thus you will not need the preallocated `res` variable

File: internal/app/collector/monitors/missing_votes_governance_monitor.go [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+	votedValidatorsSubset := make([]string, 0)
+	for _, item := range proposalVotes {
+		votedValidatorsSubset = append(votedValidatorsSubset, *item.Voter.AccountAddress)
+	}
+
+	shouldVote := m.monitoredValidators
+	res = difference(shouldVote, votedValidatorsSubset)
+
+	return res, nil
```
Comment (sotnikov-s) [inline internal/app/collector/monitors/missing_votes_governance_monitor.go:167]: would be prettier to replace with
```golang
return difference(m.monitoredValidators, votedValidatorsSubset), nil
```

File: internal/app/collector/types/test_types.go [hunk 6]
```diff
@@ -23,3 +23,9 @@ const (
 	TestMoniker        = "Test validator"
 	TestCommissionRate = 0.08
 )
+
+var (
+	MissingVotesGovernanceLookbackLimit       = 10
+	MissingVotesGovernanceAlertLimit          = 10
+	MissingVotesGovernanceMonitoredValidators = []string{"terra17kj7euq7cyedllkm5z57svvq2quxmvzq9gxfy2", "terra1xal43l3e62an3q0mky99chn9ga4j5yhnfkegx8", "terra1lph93xlfjek830h0659p0wuewu56gkjayvjucj", "terra1spvnfmgklu6ynaph007h7zszmu5sgu97k7877c", "terra1zd2dg4y3734lfywkke40dvecu8nkzd2fxp8f0v", "terra1gu9h26qeyf4etuzke3h7qq6x9rjpw4xqkprntq", "terra1xqt30x2updtqseetay5c3r7znjqfltuwc68h82", "terra1qqu376azltyc5wnsje5qgwru5mtj2yqdhj0cwl", "terra1cdzstr5qhe9wvm4ec0jmpjgprv2hpuqkx9klzn", "terra1ya8mnvt0c6rahzvgj4ezwkt49rsscn6l8k9yhv", "terra1ya8mnvt0c6rahzvgj4ezwkt49rsscn6lkeklul"}
```
Comment (sotnikov-s) [inline internal/app/collector/types/test_types.go:30]: such a long slice would look prettier if defined with an entry on a separated line

**Review by sotnikov-s — CHANGES_REQUESTED** [review]

**Comment by sotnikov-s** [comment]
is it a `chore`, not a `feat` pr?

File: internal/app/collector/monitors/missing_votes_governance_monitor.go [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+		}
+
+		for _, validatorItem := range notVotedForProposal {
+			notVotedValidators[validatorItem] += 1
+		}
+	}
+
+	for address, notVotedCount := range notVotedValidators {
+		tmpMetricVectors[MissedVotesMetric].Set(address, float64(notVotedCount))
+		tmpMetricVectors[AlertLimitMetric].Set(address, float64(m.alertLimit))
```
[the remark this answers is not in the document]
Comment (NeverHappened) [reply internal/app/collector/monitors/missing_votes_governance_monitor.go:99]: I think it's just easier to do that way, and doesn't hurt really.
It's configured from env variables, as was planned.

File: internal/app/collector/monitors/missing_votes_governance_monitor.go (hunk 5 above)
Comment (NeverHappened) [reply internal/app/collector/monitors/missing_votes_governance_monitor.go:167]: well that's just style choice)

**Comment by NeverHappened** [comment]
I don't know about feat vs chore here, need to clarify it

**Comment by sotnikov-s** [comment]
> I don't know about feat vs chore here, need to clarify it

https://www.notion.so/Policies-1c03569e77ad4c8db896c64654a19062
Github workflow, paragraph 2

**Comment by NeverHappened** [comment]
It's not just a chore, it's a new functionality, but added to the terra-monitors. That's why I though it's more like a feature than a chore

**Comment by sotnikov-s** [comment]
> It's not just a chore, it's a new functionality, but added to the terra-monitors. That's why I though it's more like a feature than a chore

I agree, but the thing is the PR is named as chore :)

File: internal/app/collector/monitors/missing_votes_governance_monitor.go [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+
+func (m *MissingVotesGovernanceMonitor) Handler(ctx context.Context) error {
+	tmpMetricVectors := make(map[MetricName]*MetricVector)
+	initMetrics(nil, m.providedMetricVectors(), nil, tmpMetricVectors)
+
+	notVotedValidators := make(map[string]uint)
+
+	proposalList, err := m.FetchLastProposals(ctx)
+	if err != nil {
+		m.logger.Errorf("Could not get proposalList. Error: %v", err)
```
[the remark this answers is not in the document]
Comment (foxpy) [reply internal/app/collector/monitors/missing_votes_governance_monitor.go:74]: If I understand correctly, the proper format string for this error would be `"Could not get proposalList: %w"`

File: docker/env/.lido_terra.testnet.env (hunk 2 above)
Comment (foxpy) [reply docker/env/.lido_terra.testnet.env:15]: And, I guess, we could make an environment variable to configure address of contract to query

File: internal/app/config/config.go [hunk 9]
```diff
@@ -50,3 +51,9 @@ type Addresses struct {
 type DelegationsDistributionConfig struct {
 	NumMedianAbsoluteDeviations int64 `envconfig:"default=3"`
 }
+
+type MissingVotesGovernanceMonitor struct {
+	LookbackLimit       int      `envconfig:"default=30"`
+	AlertLimit          int      `envconfig:"default=5"`
+	MonitoredValidators []string `envconfig:"default=terra1v5hrqlv8dqgzvy0pwzqzg0gxy899rm4kdn0jp4,terra123gn6j23lmexu0qx5qhmgxgunmjcqsx8g5ueq2,terra15zcjduavxc5mkp8qcqs9eyhwlqwdlrzy6anwpg,terra1v5hrqlv8dqgzvy0pwzqzg0gxy899rm4kdn0jp4,terra1kprce6kc08a6l03gzzh99hfpazfjeczfpd6td0,terra1c9ye54e3pzwm3e0zpdlel6pnavrj9qqvqgf7ps,terra144l7c3uph5a7h62xd8u5et3rqvj3dqtvve3he0,terra1542ek7muegmm806akl0lam5vlqlph7spfs99vq,terra1sym8gyehrdsm03vdc44rg9sflg8zeuqwfd3384,terra1khfcg09plqw84jxy5e7fj6ag4s2r9wqsg5jt4x,terra15urq2dtp9qce4fyc85m6upwm9xul30496lytpd,terra1alpf6snw2d76kkwjv3dp4l7pcl6cn9uytq89zk,terra1nwrksgv2vuadma8ygs8rhwffu2ygk4j24pxxx0,terra175hhkyxmkp8hf2zrzka7cnn7lk6mudtv4nsp2x,terra13g7z3qq6f00qww3u4mpcs3xw5jhqwraswv3q3t,terra1jkqr2vfg4krfd4zwmsf7elfj07cjuzss3qsmhm,terra15cupwhpnxhgylxa8n4ufyvux05xu864jcrrkqa"`
```
Comment (foxpy) [inline internal/app/config/config.go:58]: As @oopcode said before, this list should be queried online

**Review by foxpy — CHANGES_REQUESTED** [review]

**Review by foxpy — APPROVED** [review]

**Comment by NeverHappened** [comment]
So there is an issue that I discovered later in implementing: proposals api can give you votes only in the voting phase (see https://github.com/cosmos/cosmos-sdk/issues/10795)
And I don't know another way for now

**Pull request closed without merging**